Snail/Observable.swift

```diff
@@ -69,21 +69,18 @@ public class Observable<T>: ObservableType {
     }
 
     public func removeSubscribers() {
-        subscribersQueue.async(flags: .barrier) {
-            self.subscribers.removeAll()
-        }
+        recursiveLock.lock()
+        subscribers.removeAll()
+        recursiveLock.unlock()
     }
 
     public func removeSubscriber(subscriber: Subscriber<T>) {
-        subscribersQueue.sync {
-            guard let index = self.subscribers.enumerated().first(where: { $0.element === subscriber })?.offset else {
-                return
-            }
-
-            subscribersQueue.async(flags: .barrier) {
-                self.subscribers.remove(at: index)
-            }
+        guard let index = subscribers.enumerated().first(where: { $0.element === subscriber })?.offset else {
```

Comment:
we may want to include this in the lock

Reply:
Yes totally, its accessing the array so it should be protected as well.

We should put the lock code at the top of the function like so

```swift 
recursiveLock.lock(); defer { recursiveLock.unlock() } 
```

This will ensure that the lock is always unlocked when the scope of the function ends.